internal/workshare/workshare_waitgroup.go

```diff
@@ -0,0 +1,54 @@
+// Package workshare implements work sharing worker pool.
+package workshare
+
+import (
+	"sync"
+)
+
+// AsyncGroup launches and awaits asynchronous work through a WorkerPool.
+// It provides API designed to minimize allocations while being reasonably easy to use.
+type AsyncGroup struct {
+	wg       *sync.WaitGroup
+	requests []interface{}
+}
+
+// Wait waits for scheduled asynchronous work to complete and returns all asynchronously processed inputs.
+func (g *AsyncGroup) Wait() []interface{} {
+	if g.wg == nil {
+		return nil
+	}
+
+	g.wg.Wait()
+
+	return g.requests
+}
+
+// RunAsync starts the asynchronous work to process the provided input, the user must call Wait().
+func (g *AsyncGroup) RunAsync(w *Pool, process ProcessFunc, input interface{}) {
+	if g.wg == nil {
+		g.wg = &sync.WaitGroup{}
+	}
+
+	g.wg.Add(1)
+
+	g.requests = append(g.requests, input)
+
+	w.work <- workItem{
+		process: process,
+		input:   input,
+		wg:      g.wg,
+	}
+}
+
+// CanShareWork determines if the provided worker pool has capacity to share work.
+// If the function returns true, the use MUST call RunAsync() exactly once. This pattern avoids
+// allocations required to create asynchronous input if the worker pool is full.
```

Comment:
Any chance the expected usage pattern could be encapsulated in a helper function to simplify usage / the API?

It seems that otherwise it'd be error prone as it is fairly easy to misuse the API by calling `CanShareWork()` once and then `RunAsync()` multiple times. Or calling `CanShareWork` multiple times, but then calling `RunAsync()` fewer times, which would reduce the effective capacity of the pool.

Also, `CanShareWork` does not seem to convey the semantics of the function.

Reply:
the big problem with trying to encapsulate this is that you want to avoid allocation typically required when calling RunAsync(). Since CanShareWork() will return false most of the time, this extra if() helps avoid many allocations.